File: chatbot_api/app/routes/deploy_routes.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from dynamodb_models import User, Chat
from utils.lambda_utils import invoke_lambda
import os
import logging

logger = logging.getLogger(__name__)
deploy_bp = Blueprint('deploy', __name__)

@deploy_bp.route('/api/deploy', methods=['POST'])
@jwt_required()
def deploy_chat():
    data = request.json
    chat_id = data.get('chat_id')
    content = data.get('content')

    if not chat_id or not content:
        return jsonify({'error': 'Chat ID and content are required'}), 400

    # Get the current user from the JWT token
    current_user = get_jwt_identity()
    user = User.get_user_by_email(current_user['email'])

    if not user:
        logger.error("User not found in DynamoDB")
        return jsonify({'error': 'User not found'}), 404

    logger.info(f"Received deployment request for chat {chat_id} from user {user['id']}")

    # Define S3 file path
    s3_file_path = f"pages/{user['id']}/pages-{chat_id}/index.html"

    # Call Lambda to deploy content to S3
    lambda_payload = {
        "content": content,
        "s3_key": s3_file_path,
        "bucket_name": os.getenv('AWS_S3_BUCKET_NAME')
    }
    lambda_response = invoke_lambda("deploy_to_s3", lambda_payload)

    if 'error' in lambda_response or 's3_url' not in lambda_response:
        logger.error("Failed to deploy content via Lambda")
        return jsonify({'error': 'Failed to deploy content'}), 500

    s3_url = lambda_response['s3_url']

    # Update the chat in DynamoDB with the new page_url
    chat = Chat.get_chat_by_id(chat_id)
    if not chat:
        logger.error(f"Chat not found: {chat_id}")
        return jsonify({'error': 'Chat not found'}), 404

    Chat.update_page_url(chat_id=chat_id, page_url=s3_url)

    logger.info(f"Chat {chat_id} updated with page URL: {s3_url}")
    return jsonify({'page_url': s3_url}), 200

File: node_backend/routes/deployRoutes.js
import express from 'express';
import jwt from 'jsonwebtoken';
import { User, Chat } from '../models/dynamodbModels.js';
import getSecrets from '../utils/getSecrets.js';
import { deployToS3 } from '../utils/deployToS3.js';

const deployRouter = express.Router();
const secrets = await getSecrets();
const JWT_SECRET_KEY = secrets.JWT_SECRET_KEY;

// Middleware to verify JWT token and extract user identity
const authenticate = (req, res, next) => {
    const token = req.headers.authorization?.split(' ')[1];
    if (!token) return res.status(401).json({ error: 'Unauthorized' });

    try {
        const decoded = jwt.verify(token, JWT_SECRET_KEY);
        req.user = decoded; // Attach user data to the request
        next();
    } catch (error) {
        return res.status(401).json({ error: 'Invalid token' });
    }
};

// Deploy Chat Route
deployRouter.post('/', authenticate, async (req, res) => {
    const { chat_id: chatId, content } = req.body;

    if (!chatId || !content) {
        return res.status(400).json({ error: 'Chat ID and content are required' });
    }

    const currentUserEmail = req.user.email;

    // Fetch the user from the database
    const user = await User.getUserByEmail(currentUserEmail);
    if (!user) {
        console.error("User not found in the database");
        return res.status(404).json({ error: 'User not found' });
    }

    console.log(`Received deployment request for chat ${chatId} from user ${user.id}`);

    const s3FilePath = `pages/${user.id}/pages-${chatId}/index.html`;
    const s3Url = await deployToS3(content, secrets.AWS_S3_OUTPUT_BUCKET_NAME, s3FilePath);

    if (!s3Url || s3Url === "") {
        console.error("Failed to deploy content");
        return res.status(500).json({ error: 'Failed to deploy content' });
    }

    // Update the chat in the database with the new page URL
    const chat = await Chat.getChatById(chatId);
    if (!chat) {
        console.error(`Chat not found: ${chatId}`);
        return res.status(404).json({ error: 'Chat not found' });
    }

    await Chat.updatePageUrl(chatId, s3Url);

    console.log(`Chat ${chatId} updated with page URL: ${s3Url}`);
    return res.status(200).json({ page_url: s3Url });
});

export default deployRouter;
